### PR TITLE
Assign all sources/shards, even if this require exceeding the indexer

### LINKED ISSUE #4290: Allow over assigning
Related to #4055 (somewhat)

Not assigning shard is not a good option. It will evenutally stall the WAL.
We shoudl log an error and assign the remaining pipelines.

### PR DESCRIPTION
**Pull request by fulmicoton**
capacities.

We do that simply by scaling the indexer capacities. First we scale the indexer capacities so that the sum of their capacities exceeds the total load of the source to assign.

If the packing algorithm fails, we inflates the capacities again by 10% and retry.

Closes #4290

### CODE REVIEW

File: quickwit/quickwit-control-plane/src/indexing_scheduler/mod.rs [hunk 1]
```diff
@@ -780,37 +780,7 @@ mod tests {
...
-                        }
-                        SourceToScheduleType::NonSharded {
-                            num_pipelines: _ ,
-                            load_per_pipeline,
-                        } => {
-                            load_in_node += load_per_pipeline.get();
-                        }
-                    }
-                }
-                assert!(load_in_node <= indexer_max_loads.get(node_id).unwrap().cpu_millis());
```
Comment (fulmicoton) [inline quickwit/quickwit-control-plane/src/indexing_scheduler/mod.rs:812]: After this issue, this assert is actually invalid so I removed it.
I added an assert checking that the physical plan matches the solution.

File: quickwit/quickwit-control-plane/src/indexing_scheduler/scheduling/mod.rs [hunk 2]
```diff
@@ -124,7 +125,19 @@ fn convert_physical_plan_to_solution(
...
                 if let Some(source_ord) = id_to_ord_map.source_ord(&source_uid) {
-                    indexer_assignment.add_shards(source_ord, indexing_task.shard_ids.len() as u32);
+                    let source_type = &sources[source_ord as usize].source_type;
+                    match source_type {
+                        SourceToScheduleType::Sharded { .. } => {
+                            indexer_assignment.add_shards(source_ord, indexing_task.shard_ids.len() as u32);
+                        }
+                        SourceToScheduleType::NonSharded { .. } => {
+                            // For non-sharded sources like Kafka, one pipeline = one shard in the solutions
+                            indexer_assignment.add_shards(source_ord, 1);
```
Comment (fulmicoton) [inline quickwit/quickwit-control-plane/src/indexing_scheduler/scheduling/mod.rs:135]: the conversion for non sharded sources (kafka) was incorrect.

File: quickwit/quickwit-control-plane/src/indexing_scheduler/scheduling/mod.rs [hunk 3]
```diff
@@ -124,7 +125,21 @@ fn convert_physical_plan_to_solution(
                     source_id: indexing_task.source_id.clone(),
                 };
                 if let Some(source_ord) = id_to_ord_map.source_ord(&source_uid) {
-                    indexer_assignment.add_shards(source_ord, indexing_task.shard_ids.len() as u32);
+                    let source_type = &source_ord_map.get(&source_ord).unwrap().source_type;
```
Comment (fmassot) [inline quickwit/quickwit-control-plane/src/indexing_scheduler/scheduling/mod.rs:128]: I'm always afraid of those unwraps, I'm unsure if a comment is worth it on the function or maybe replace unwrap by expect("")

File: quickwit/quickwit-control-plane/src/indexing_scheduler/scheduling/scheduling_logic.rs [hunk 4]
```diff
@@ -195,84 +222,85 @@ fn assert_enforce_nodes_cpu_capacity_post_condition(
...
-        );
-        // We haven't been able to place this source entirely.
-        if num_shards_unassigned != 0 {
-            unassignable_shards.insert(source.source_ord, num_shards_unassigned);
+
+    // Thanks to the call to `inflate_node_capacities_if_necessary`,
+    // we are certain that even on our first attempt, the total capacity of the indexer exceeds 120%
+    // of the partial solution.
+    //
+    // 1.2^30 > 240.
```
Comment (fmassot) [inline quickwit/quickwit-control-plane/src/indexing_scheduler/scheduling/scheduling_logic.rs:244]: 1.2**30 ~ 237.37

File: quickwit/quickwit-control-plane/src/indexing_scheduler/scheduling/mod.rs (hunk 2 above)
Comment (fmassot) [reply quickwit/quickwit-control-plane/src/indexing_scheduler/scheduling/mod.rs:137]: In the NonSharded source, you have a `num_pipelines`, should we not use it?

Comment (fmassot) [reply quickwit/quickwit-control-plane/src/indexing_scheduler/scheduling/mod.rs:137]: forget my comment, we are iterating on indexer task here.

**Review by fmassot — APPROVED** [review]
Approved. Do you know if our tests are currently sufficient? Wondering if adding protesting is worth it

**Pull request merged**